src/crimson/os/seastore/journal/circular_bounded_journal.cc

```diff
@@ -0,0 +1,570 @@
+      "Invalid error write_header"
+    }
+  );
+}
+
+CircularBoundedJournal::open_for_write_ret
+CircularBoundedJournal::open_device_read_header(rbm_abs_addr start)
+{
+  LOG_PREFIX(CircularBoundedJournal::open_for_write);
+  ceph_assert(!init);
+  return _open_device(path
+  ).safe_then([this, start, FNAME]() {
+    return read_header(start
+    ).handle_error(
+      open_for_write_ertr::pass_further{},
+      crimson::ct_error::assert_all{
+	"Invalid error read_header"
+    }).safe_then([this, start, FNAME](auto p) mutable {
+      auto &[head, bl] = *p;
+      header = head;
+      DEBUG("header : {}", header);
+      paddr_t paddr = convert_abs_addr_to_paddr(
+	get_written_to(),
+	header.device_id);
+      start_dev_addr = start;
+      init = true;
+      return open_for_write_ret(
+	open_for_write_ertr::ready_future_marker{},
+	journal_seq_t{
+	  cur_segment_seq,
+	  paddr
+	});
+    });
+  }).handle_error(
+    open_for_write_ertr::pass_further{},
+    crimson::ct_error::assert_all{
+      "Invalid error _open_device"
+  });
+}
+
+CircularBoundedJournal::write_ertr::future<> CircularBoundedJournal::append_record(
+  ceph::bufferlist bl,
+  rbm_abs_addr addr)
+{
+  LOG_PREFIX(CircularBoundedJournal::append_record);
+  std::vector<std::pair<rbm_abs_addr, bufferlist>> writes;
+  if (addr + bl.length() <= get_journal_end()) {
+    writes.push_back(std::make_pair(addr, bl));
+  } else {
+    // write remaining data---in this case,
+    // data is splited into two parts before due to the end of CircularBoundedJournal.
+    // the following code is to write the second part
+    bufferlist first_write, next_write;
+    first_write.substr_of(bl, 0, get_journal_end() - addr);
+    writes.push_back(std::make_pair(addr, first_write));
+    next_write.substr_of(
+      bl, first_write.length(), bl.length() - first_write.length());
+    writes.push_back(std::make_pair(get_start_addr(), next_write));
+  }
+
+  return seastar::do_with(
+    std::move(bl),
+    [this, writes=move(writes), FNAME](auto& bl) mutable
+  {
+    DEBUG("original bl length {}", bl.length());
+    return write_ertr::parallel_for_each(
+      writes,
+      [this, FNAME](auto& p) mutable
+    {
+      DEBUG(
+	"append_record: offset {}, length {}",
+	p.first,
+	p.second.length());
+      return device_write_bl(p.first, p.second 
+      ).handle_error(
+	write_ertr::pass_further{},
+	crimson::ct_error::assert_all{ "Invalid error device->write" }
+      ).safe_then([]() {
+	return write_ertr::now();
+      });
+    });
+  });
+}
+
+CircularBoundedJournal::submit_record_ret CircularBoundedJournal::submit_record(
+  record_t &&record,
+  OrderingHandle &handle)
+{
+  LOG_PREFIX(CircularBoundedJournal::submit_record);
+  assert(write_pipeline);
+  auto r_size = record_group_size_t(record.size, get_block_size());
+  auto encoded_size = r_size.get_encoded_length();
+  if (get_written_to() +
+      ceph::encoded_sizeof_bounded<record_group_header_t>() > get_journal_end()) {
```

Comment:
(this got missed from yesterday)



This check/behavior belongs in append_record, it's a special case of generally needing to split a record across the end and beginning of the journal.

Reply:
The purpose of this line is to set written_to to the start address of CBjournal because of there is not enough space (< records_group_header_t) between written_to---address to write a record---and the end of CBjournal. So, the record will be append to the beginning of CBjournal. 
Note that this line does not intend to split a record the end and beginning of the CBjournal.

Reply:
Ok.